### DISCUSSION TITLE
Hexagon Filter?

### DISCUSSION THREAD

**Discussion by 60horsesinmyherd** [ImageMagick/ImageMagick#7417] [Help] [answered]
### ImageMagick version

7.1.1-33

### Operating system, version and so on

Windows 10

### Description

I'm very green with this software (and the command-line in general), so any help would be really appreciated.

I need to filter an image like this:
![123](https://github.com/ImageMagick/ImageMagick/assets/173410406/c4d6bf90-8c8e-4777-b8a8-60fccc65f954)

The catch is that I need the hexagons to be exactly these dimensions:
![example](https://github.com/ImageMagick/ImageMagick/assets/173410406/5074d6c2-3d64-431d-a05a-923e395414cc)

Such that the hexes in the filtered image would be arranged like this:
![zop](https://github.com/ImageMagick/ImageMagick/assets/173410406/568dddd3-6c11-4d91-9a42-43b9f8a212c3)

Is this possible? If so, how might I achieve it?

**Comment by fmw42** [comment]
I have a bash shell script for Imagemagick that will do hexagons but may not be fully compliant with your requirements.  See stainedglass at http://www.fmwconcepts.com/imagemagick/stainedglass/index.php.

Input:

![birdofparadise](https://github.com/ImageMagick/ImageMagick/assets/13458218/0a9bbbd2-5bc5-43de-b968-eb82b8db10ac)

Two results where I specify the "size" of the hexagons as either 37 or 31.  Let me know if either are what you desire.

37:

![birdofparadise_hex_37](https://github.com/ImageMagick/ImageMagick/assets/13458218/a386a277-081f-4a3e-a5ab-11cabd6c8093)

31:

![birdofparadise_hex_31](https://github.com/ImageMagick/ImageMagick/assets/13458218/5a569955-f70f-4f06-84f2-1405cd422603)


It looks to me like my hexagons are 90 deg rotated from the way you want them, unfortunately. So one way to make them the other orientation would be to rotate the image 90 deg before the stainedglass effect was done and then rotate back by -90 deg afterwards.

Here is the result of doing that on the 37 size:

![x3](https://github.com/ImageMagick/ImageMagick/assets/13458218/27670821-6521-4766-9da2-7bdc301dbe54)

**Reply by 60horsesinmyherd** [reply]
> I have a bash shell script for Imagemagick that will do hexagons but may not be fully compliant with your requirements. See stainedglass at http://www.fmwconcepts.com/imagemagick/stainedglass/index.php.

Functionality wise this is perfect, I just need the hex to be the right dimensions and it would be exactly what I need :(. 
![pain of a](https://github.com/ImageMagick/ImageMagick/assets/173410406/cbb0b8e7-1018-40ec-9312-930ac7f9c79e)

Could my hex be added in as a parameter for the script?

**Reply by fmw42** [reply]
You should get what you want using the technique I showed about rotating 90 deg and back and adjusting the size argument to get the dimensions you want.  There should be no need for any other parameter. 

I do not understand your diagram above.

**Reply by 60horsesinmyherd** [reply]
> I do not understand your diagram above.

Sorry, I could've made it more clear. 
![image](https://github.com/ImageMagick/ImageMagick/assets/173410406/4c98d7e7-6a82-4561-b8e2-7fae4b79021c)
This is the size 31 hexagon you provided but flipped. From top to bottom it's perfectly aligned, but from side to side it isn't. In the original diagram I was trying to show that the ratios of the top to sides were different between our hexes (it would've been much more clear if I had used the 31 hex in the first place). 
Is there a way to alter side profile of the hexes in your script?

**Reply by fmw42** [reply]
Right now, no, only by the size argument?  What is your ratio and what do you get from mine?  Can you not change the size argument to get what you want?  Why is that so critical?  I am using the ratios of a normal hexagon.  I believe that I coded it according to the equivalent of https://en.wikipedia.org/wiki/Hexagon.  See also https://math.stackexchange.com/questions/2688763/formula-for-the-points-in-a-grid-of-hexagons

**Reply by fmw42** [reply]
Are you asking about the extra rotations?  If so, I can try to add that to the script. But changing the horizontal to vertical ratio is not easy. My code does not have an argument for that right now. It only produces the "regular" hexagon ratio of horizontal to vertical as defined in the wikipedia article, so all sides are defined so as to have the same interior angle (120 deg) or angle between vertices (60 deg). See diagram in the wikipedia link above.

**Reply by 60horsesinmyherd** [reply]
> Why is that so critical? 

It's integral to my workflow unfortunately. It might be a bit more clear if I give an example of my end goal. I'm trying to make a hexmap just like the following: 
<img src="https://github.com/ImageMagick/ImageMagick/assets/173410406/67a6760f-7dcb-44be-8393-773441467b7b" width="400" height="350" />
The sprites I'm using will only work if hexes fit together like the above.
I think what I'm getting from your responses is that that might not be feasible. I need the hexes to be exactly the same dimensions and side profile/width as the purple hex. There can't be any deviation or the entire thing winds up eschew.

**Reply by fmw42** [reply]
My hexagons fit perfectly also.  Your hexagons may not conform to "regular" hexagons.

I measure roughly 36 wide by 30 tall hexagons on your image, which could be 37 x 31 and that is exactly the size of my hexagons when rotated 90. Or you could specify 36 or 30 and get the shape you have.  How did you get your measurements?

**Reply by fmw42** [reply]
I have put in a rotation argument in my scripts, which seems to work. But currently the script works in IM 6, but fails in IM 7. There is a bug that I have reported at https://github.com/ImageMagick/ImageMagick/issues/7422.  So I have to wait until that gets fixed to test my script for IM 7.

**Reply by 60horsesinmyherd** [reply]
> I measure roughly 36 wide by 30 tall hexagons on your image, which could be 37 x 31 and that is exactly the size of my hexagons when rotated 90. Or you could specify 36 or 30 and get the shape you have. How did you get your measurements?

Just to be clear, you're talking about the image you rotated in your original reply?

Assuming yes, here's how I'm measuring it:
![image](https://github.com/ImageMagick/ImageMagick/assets/173410406/cd134a0c-4565-4cea-adf7-7975b908f05b)
^ This is a hex extracted from your original reply.
![image](https://github.com/ImageMagick/ImageMagick/assets/173410406/60c10117-b97b-49c4-b3c2-9fed33a48cd2)
^ This is the shape I would need it to be. 

You're saying if I adjust the size parameter in your script, I can achieve this:
![image](https://github.com/ImageMagick/ImageMagick/assets/173410406/551a61a3-16fe-4f0a-8125-d5fcf883d549)
?
I apologize if I'm being obtuse.

**Reply by fmw42** [reply]
No I was measuring in the hex grid image that you showed above.

Yes, if you adjust the size it should scale to your size.  You have to choose carefully, using the rotate 90 and rotate -90 (right now as pre and post processing) to get the hexagon horizontal. But you need to measure the height rather than the width when doing these rotations, then check the width. Adjust as needed. So measure your desired height and try that value for the size argument. But rotate your image 90 before the script and then rotate it the opposite after the script, i.e. -90.

I do not know about the 1 px vs 2 px wide vertices.  That may have to do with just how Imagemagick draws the hexagons.

**Reply by fmw42** [reply]
I suspect the size of the vertices has to do with how Imagemagick is drawing the hexagons.  What I do is compute a grid of vertices according to https://www.bartleby.com/solution-answer/chapter-138-problem-11we-mcdougal-littell-jurgensen-geometry-student-edition-geometry-5th-edition/9780395977279/fed2d33d-4d88-46df-9322-145680ee7a09%23google_vignette.  I draw the points by drawing alternating rows with different spacing and offsets. This draws the following:

![tmpC](https://github.com/ImageMagick/ImageMagick/assets/13458218/b6b47956-7ca6-4086-b990-34aebb464150)

Then I get the color at each white point:

[tmpSC.txt.zip](https://github.com/user-attachments/files/15942183/tmpSC.txt.zip)

Then I use -sparsecolor voronoi to get the closest color to grid colors for every other pixel that is not on the grid and draws the hexagon.  See https://imagemagick.org/Usage/canvas/#voronoi

![birdofparadise_stain_hex_t0_s31_6](https://github.com/ImageMagick/ImageMagick/assets/13458218/2dd02508-4fbf-47bb-b46e-0e840c137601)

**Reply by 60horsesinmyherd** [reply]
> Then I get the color at each white point:

Interesting. Just as an idea, would it be possible to develop a method using this to fill only within the bounds of a mask?

![image](https://github.com/ImageMagick/ImageMagick/assets/173410406/5964f266-ff9e-4ee4-b62a-5a814dd1623a)

Where the white portions would be filled with whatever color pixel is at its center.

**Reply by fmw42** [reply]
No, not easily. How would you create the mask?  Then the only way I know would be to flood fill each cell, which would be tedious.  Then what about the black lines?  They would not be filled.

But if you want to overlay a hexagon grid of cells onto some image like in your game link, then use the hexagon: pattern and scale to the size you need, then tile it out. Then composite over your image.

See

https://imagemagick.org/script/formats.php#builtin-patterns
https://imagemagick.org/Usage/canvas/#tile (keep reading down for more tiling options)
https://imagemagick.org/Usage/canvas/#pattern

**Reply by fmw42** [reply]
P.S. better yet, you can create your own hexagon outline with https://imagemagick.org/Usage/draw/#paths then tile it out.  Use the dimensions for the vertices from the link I provided above.

**Reply by 60horsesinmyherd** [reply]
> No, not easily. How would you create the mask? Then the only way I know would be to flood fill each cell, which would be tedious. 

I could create the mask in photoshop using the pattern tool (that's how I made the example above). That part would be more or less instant. 
I was moreso trying to ask if the software would support using a mask that I generated externally. As in, using such a mask to define where a given color that's been sampled fills in.

>Then what about the black lines? They would not be filled.

That's actually what I want too.

**Reply by fmw42** [reply]
Please explain in more detail. Do you need the black lines filled?  What is your ultimate goal?  If it is to make an image like the game one you showed, then you need the black lines!  How would you use your mask to achieve such goal.  You cannot just fill each cell of your mask with image color and fill also the black lines, if you want to use it to do something like the my stainedglass script.

However, in reviewing my code, I could add a scaleX and scaleY in place of the simple scale, but I cannot guarantee that it would tile out without gaps. I would have to code and test it.  

But have you tried to use my script and adjust the scale to see if it measures the way you want?

**Reply by 60horsesinmyherd** [reply]
> But have you tried to use my script and adjust the scale to see if it measures the way you want?

Not yet. I've been having trouble getting ImageMagick to work properly on Windows. I'm now trying to get it working through the Linux Bash Shell but that's proving to be equally challenging. I'm not much of a Linux user at all to be frank.

> Please explain in more detail.

Here's an infographic that should hopefully better explain my idea/thinking. 
If the following is possible, it would give me the exact results I needed.

![1](https://github.com/ImageMagick/ImageMagick/assets/173410406/8d8c82d4-f8d7-49ab-91a1-f4833592e739)
![2 2](https://github.com/ImageMagick/ImageMagick/assets/173410406/af25f9ae-ce3d-470d-b331-9c43caaf324b)
![2 3](https://github.com/ImageMagick/ImageMagick/assets/173410406/c932ff61-d55d-48cc-98d5-8012b521dd33)
![3](https://github.com/ImageMagick/ImageMagick/assets/173410406/69327f61-9e7c-4696-bcf9-e70548b7d69f)
![5 1](https://github.com/ImageMagick/ImageMagick/assets/173410406/f9bf41f0-c3b0-4118-81b4-8b5e71ca6b7f)
![4poin](https://github.com/ImageMagick/ImageMagick/assets/173410406/4c935efd-fb96-4a36-a933-1134a3d04005)
![4point](https://github.com/ImageMagick/ImageMagick/assets/173410406/239d82b4-aa64-4707-b0d6-ad91725ac533)
![6](https://github.com/ImageMagick/ImageMagick/assets/173410406/da88dab4-6ec7-45db-8178-05bb0a6b92c3)

**Reply by fmw42** [reply]
Most is easy except filling each hexagon with the color from the center points. IM has a flood fill that could be used, but it only does one cell. So one would have to loop over each center point and do the flood fill of your empty cells as long as the center color is different from the hexagon outlines. This could take some time. It would not be quick. It might be quicker to collect all command for -draw with floodfill into one draw command rather than loop over each cell doing either -draw with floodfill or -floodfill.

The other issue is computing the average color inside the cell.  It is easy to get the color at the center points if you provide the point array image, but not the average color. However, the center color may not be the dominant color in the hexagon.

Provide me a small example of your image, the hexagon cells image (perhaps white outline and black fill), and the points image (white with black background). I will see what I can do with that example test.

**Reply by fmw42** [reply]
Please explain the the game image vs the first image you provided.  I do not understand where the game image comes into this. I thought you wanted to use the first image and convert that to hexagon cells.

**Reply by fmw42** [reply]
If you can provide the points as a list of coordinate pairs x,y, then that might be even better. But if not, then the image of points is acceptable.

**Reply by 60horsesinmyherd** [reply]
> Please explain the the game image vs the first image you provided. I do not understand where the game image comes into this. I thought you wanted to use the first image and convert that to hexagon cells.

The image in my original post was just an abstract image I used to illustrate the operation I was trying to achieve. 
The game image (which I didn't create) is what I'd be using that operation to try to achieve.

> However, the center color may not be the dominant color in the hexagon.

That wouldn't be a huge problem. If just using the center color is easier/faster I wouldn't have any complaints.
Here's the images you requested.
![base](https://github.com/ImageMagick/ImageMagick/assets/173410406/ba80a896-f67d-4c9c-8d3b-6bc37692b658)
![poitns](https://github.com/ImageMagick/ImageMagick/assets/173410406/271f5f60-a497-4ebf-a99a-e6b3759089c7)
![mask](https://github.com/ImageMagick/ImageMagick/assets/173410406/e4a42d82-1242-4b29-8d29-16c03b0d7a9b)

**Reply by fmw42** [reply]
Note that the partial hexagons along the border at the top, right and bottom are not going to be filled here since there are no center points for them. The ones on the left seems to have center points since the hexagons are more than half showing.

**Reply by 60horsesinmyherd** [reply]
> Note that the partial hexagons along the border at the top, right and bottom are not going to be filled here since there are no center points for them. The ones on the left seems to have center points since the hexagons are more than half showing.

Yes, I noticed that. Since this is just an example it should be okay. In actual use I can just add some extra blank pixels around the edges and then crop as necessary.

**Comment by fmw42** [comment]
Here is a script for bash unix OS.  It is faster than I thought but still not quick at about 8-10 sec.

```
# set images
infile="image.png"
centfile="centers.png"
hexfile="hexagons.png"

# set desired color for hexagon outlines -- provided it is not a color in the infile
bgcolor="red"

# convert centfile to white on transparent background, then use sparse-color: to get coords and colors from infile
list=`magick "$infile" \( "$centfile" -transparent black \) -alpha off -compose copy_opacity -composite sparse-color: | tr " " "\n"`
#echo "$list"

# convert hexagon outlines to bgcolor and then floodfill each hexagon with colors from list at seed point from coords in list
inname=`magick "$infile" -format "%t" info:`
magick "$hexfile" \( +clone -fill red -colorize 100 \) +swap -compose multiply -composite tmp.miff
for item in $list; do
coords=`echo $item | cut -d, -f1-2`
color=`echo $item | cut -d, -f3-6`
#echo "coords=$coords; color=$color;"
echo "fill $color color $coords floodfill"
magick tmp.miff-draw "fill $color color $coords floodfill" tmp.miff
done
magick tmp.miff "${inname}_hexed".png
rm -f tmp.miff

```

Result Is:

![image_hexed](https://github.com/ImageMagick/ImageMagick/assets/13458218/3a8c2b0d-f10a-4849-ae51-91a260a3fb5a)

**Comment by snibgo** [comment]
@fmw42: Your script runs IM at every hexagon, reading and writing to tmp.png. Reading and writing to tmp.miff or tmp.mpc would be faster.

Better still, run `magick` just once, without reading and writing a temporary file. The loop would write to an IM script file, that could end with:
```
-write hexed.png
-exit
```
Then execute the script with:
```
magick -script myscript.scr
```

**Reply by fmw42** [reply]
@snibgo Yes, those would help. I changed the script to use tmp.miff.  I can't make it one call to IM as I need to change the color of the white hexagon outlines. Even if I kept the outlines as white, I would then be writing each loop as the file type of the input, which in this case is png.

I tried accumulating the -draw command arguments for the looping but found that -draw did not like to flood fill more than one region. It complained about doing two.  I have yet to have time to check about multiple -floodfill commands chained or multiple -draw commands chained to see if either of those will make it faster.

However, the user is on Windows. So he will need some kind Windows expert to convert this idea to that OS for him.

**Reply by fmw42** [reply]
Update:  I made a typo in my test of accumulating two floodfill arguments in the draw command. It does work with two.  So I need to go back and see why my loop messed up with doing all of them. Perhaps the command was too long.

**Comment by fmw42** [comment]
Here is a faster version by accumulating all the arguments for the the flood filling into one -draw command. The runs in about 1 sec.  However, it might run into problems with a vary large image and too many characters in the accumulated argument for -draw.  This approach might be especially problematic on Windows if there is still a 256 character limit.

```
# set images
infile="image.png"
centfile="centers.png"
hexfile="hexagons.png"

# set desired color for hexagon outlines -- provided it is not a color in the infile
bgcolor="red"

# convert centfile to white on transparent background, then use sparse-color: to get coords and colors from infile
list=`magick "$infile" \( "$centfile" -transparent black \) -alpha off -compose copy_opacity -composite sparse-color: | tr " " "\n"`
#echo "$list"

# convert hexagon outlines to bgcolor and then floodfill each hexagon with colors from list at seed point from coords in list
inname=`magick "$infile" -format "%t" info:`
filling=""
for item in $list; do
coords=`echo $item | cut -d, -f1-2`
color=`echo $item | cut -d, -f3-6`
filling="$filling fill $color color $coords floodfill"
done
#echo $filling
magick "$hexfile" \( +clone -fill red -colorize 100 \) +swap -compose multiply -composite -draw "$filling" -alpha off "${inname}_hexed".png

```

![image_hexed](https://github.com/ImageMagick/ImageMagick/assets/13458218/6fc179d1-084d-469e-9167-9c9c7755616a)

**Comment by fmw42** [comment]
@60horsesinmyherd No comment from you after all this work?

**Reply by 60horsesinmyherd** [reply]
> @60horsesinmyherd No comment from you after all this work?

I was away from my desktop yesterday. I think this is exactly what I need. Thank you. Once I get ImageMagick working properly I'll give it a whirl.

**Comment by snibgo** [comment]
> This approach might be especially problematic on Windows if there is still a 256 character limit.

The Windows limit on an environment variable is around 4400 characters. If that is a problem, the script could be modified to echo each `fill $color color $coords floodfill` to a file or a pipe, which `-draw` would read with `@`. I can help with that, if needed.

I am working on a general polygon pixelation method. It takes an input image, and a same-sized image of polygons (not necessarily regular, or even a repeated pattern), where each polygon is a different colour. The output image is the same size. Each output pixel is set to the average colour of all the input pixels that are in the corresponding polygon.

I am currently writing this in C.

**Reply by 60horsesinmyherd** [reply]
> The Windows limit on an environment variable is around 4400 characters. If that is a problem, the script could be modified to echo each `fill $color color $coords floodfill` to a file or a pipe, which `-draw` would read with `@`. I can help with that, if needed.

Apologies if my question is a little redundant, but if you did do that, would it resolve the potential issues with large image sizes that @fmw42 mentioned? 
Some of the maps I'm trying to make can be quite large (upwards of 9000x9000 pixels), so if editing the script in this way could accommodate for that it could save a lot of headache for me down the road.

**Reply by fmw42** [reply]
@snibgo. Just curious how do you determine the average color for a given polygon from the input image?

@60horsesinmyherd. Snibgo is very clever and knows C programming and Windows. So his program will be much more efficient than my script.  Your 9000x9000 size image with the same size hexagons will exceed the 4400 character limit of Windows and so my script will not work unless a temporary text file was written holding the command and then read back where needed in the Imagemagick command to do the coloring. So a C program for Imagemagick will be faster and can avoid the character limit in some way if not by writing to a temporary text file.  Snibgo will likely have a clever solution and resulting code and Imagemagick processing function.

**Comment by snibgo** [comment]
Here is a translation of fmw42's most recent script into Windows BAT. It uses a temporary file `drawStuff.scr` and there is no limit to its size. 
```
:: set images
set infile=imgForHex.png
set centfile=centForHex.png
set hexfile=hexForHex.png
set drawfile=drawStuff.scr

:: set desired color for hexagon outlines -- provided it is not a color in the infile
set bgcolor="red"

for /F "usebackq" %%L in (`magick "%infile%" -format "inname=%%t" info:`) do set %%L


:: convert centfile to white on transparent background, then use sparse-color: to get coords and colors from infile

echo off
(
  for /F "usebackq tokens=1-6 delims=," %%A in (`magick ^
    "%infile%" ^
    ^( "%centfile%" -transparent black ^) ^
    -alpha off ^
    -compose copy_opacity -composite ^
    sparse-color: ^| tr " " "\n"`) do (
    echo fill %%C,%%D,%%E^) color %%A,%%B floodfill
  )
) >%drawfile%

echo on

type %drawfile%

magick ^
  "%hexfile%" ^
  ( +clone -fill red -colorize 100 ) ^
  +swap ^
  -compose multiply -composite ^
  -draw @%drawfile% ^
  -alpha off ^
  "%inname%_hexed.png"

del %drawfile%
```

>  Just curious how do you determine the average color for a given polygon from the input image?

In principle, like this:
```
for each polygon {
  for each pixel in that polygon {
    add the colours
  }
  divide the sum of the colours by the number of pixels
}
```
Visiting pixels that are in a polygon isn't trivial, so the algorithm doesn't work like that. Instead, the unique colours define the polygons. Then:
```
For each pixel in the polygon image, find which polygon is belongs to.

For each pixel in the main input image, add to the colour in another location, and increment the divisor for that polygon.

For each polygon, divide the sum of the colours by the divisor.
```
Then we can:
```
For each output pixel, paint it with the mean colour of the location's polygon.
```

This is different to fmw42's script above. That script takes the central pixel of each hexagon, and makes the entire hexagon that colour. If a suitable blur is first done on the input image, then the central pixel of each hexagon will approximate the mean colour of all the pixels in the hexagon.

My method calculates the accurate mean of the pixels in each hexagon, or whatever the shapes are.

When calculating the mean, I'll provide the facility to treat alpha like any other channel, or do the usual trick of multiplying each colour by alpha and summing the results, then dividing by the sum of the alphas instead of simply the number of pixels.

I doubt that my C code will be faster than fmw42's script. That isn't my goal. Instead, I aim for accuracy and generality.

If greater speed is required, C code could be written just for hexagons.

**Reply by 60horsesinmyherd** [reply]
I just tried this and it runs! But this is the result for some reason:
![imgForHex_hexed](https://github.com/ImageMagick/ImageMagick/assets/173410406/5c887a00-b97c-4af3-9b74-c2a94a470008)

I've tried inverting the mask, inverting the center pixels map, removing the background on the center pixel map, and a few other things, all to no avail. Could it be a problem with my ImageMagick install?
Here's the images I'm using (scaled down a bit for readability):
<img src="https://github.com/ImageMagick/ImageMagick/assets/173410406/7354ab0b-c8da-4ccc-b266-35c4a427e23f" width="640" height="400" />

<img src="https://github.com/ImageMagick/ImageMagick/assets/173410406/21de05b4-d1a6-490d-b22d-85cca2725624" width="640" height="400" />

<img src="https://github.com/ImageMagick/ImageMagick/assets/173410406/2f98d83d-1116-4eb9-8edd-457b8a20ce2e" width="640" height="400" />

**Comment by fmw42** [comment]
@snibgo > For each pixel in the polygon image, find which polygon is belongs to.

That is the part I do not understand.  How are you determining in which polygon a pixel in the input image belongs using the multi-hexagons outline image.  Are you doing a pixel search in all directions until you hit a white pixel?

Thanks

**Comment by snibgo** [comment]
> I just tried this and it runs! But this is the result for some reason:

Using IM v7.1.1-20 on Windows 11, with your three input images, my script gives this result for me:

![hexImg2_hexed](https://github.com/ImageMagick/ImageMagick/assets/33812368/ea3604ee-9816-4068-9e25-b413a8784d23)

> How are you determining in which polygon a pixel in the input image belongs using the multi-hexagons outline image.

In that image, the hexagons must be different colours. It doesn't matter what the colours are, provided they are different. (This can be enforced by using connected components.) My code finds the unique colours in that image. Then, for each pixel in that image, my code finds which of the unique colours the pixel is, so that determines which hexagon it is in. Searching the unique colours is fairly fast because the pixel is likely to be in the same hexagon as the previous pixel.

**Reply by 60horsesinmyherd** [reply]
> Using IM v7.1.1-20 on Windows 11, with your three input images, my script gives this result for me:

I'm on Windows 10 and using IM v7.1.1-34. Could that be why it's not working for me?

**EDIT**: I grabbed IM v7.1.1-20 through the Wayback Machine and then tried again, but the output on my end remains black.
Here's a screenshot of the directory incase that's relevant. 
![image](https://github.com/ImageMagick/ImageMagick/assets/173410406/6e7213a8-3716-4da0-bbfe-b87ff4bf2657)

And a screenshot of the .bat in Notepad++
<img src="https://github.com/ImageMagick/ImageMagick/assets/173410406/7c7a981c-66a3-4d1a-a80a-1358ab0e5366" width="640" height="400" />

**Comment by snibgo** [comment]
Please remove (or comment out) the bottom line:
```
del %drawfile%
```
... then run the script, and then paste here the contents of `drawStuff.scr`, which is a text file.

**Reply by 60horsesinmyherd** [reply]
> ... then run the script, and then paste here the contents of `drawStuff.scr`, which is a text file.

![image](https://github.com/ImageMagick/ImageMagick/assets/173410406/130b3028-cf98-4c48-b5ae-a9d8f50cb532)
![image](https://github.com/ImageMagick/ImageMagick/assets/173410406/48fbf950-bb4f-4c45-a0bd-bc3a56d2eba1)

It appears to be blank.

**Comment by snibgo** [comment]
Do you have the program `tr`? What happens when you type `tr` at the console?

If you don't have it, then I suggest you get it. Or you could do the job with `sed`.

`tr` and `sed` and other common utilities are available for Windows from https://cygwin.com/, and probably many other places.

**Reply by 60horsesinmyherd** [reply]
> Do you have the program `tr`? What happens when you type `tr` at the console?

![image](https://github.com/ImageMagick/ImageMagick/assets/173410406/e5a24c04-8a3a-41c9-9d86-3846da93e92a)
It would seem that I don't. Let me try to figure out how to get that command working.

**Reply by 60horsesinmyherd** [reply]
> `tr` and `sed` and other common utilities are available for Windows from https://cygwin.com/, and probably many other places.

What packages contain these commands? I downloaded and installed Cygwin via setup-x86_64.exe, but I'm still getting tr not recognized.

**Reply by 60horsesinmyherd** [reply]
Nevermind! I fixed it. If I drag and drop the .bat into a cygwin terminal it runs and produces this:
![imgForHex_hexed](https://github.com/ImageMagick/ImageMagick/assets/173410406/87aa8fc8-9487-4c41-901f-c8ad0411721f)
Provided it's in this directory:
![image](https://github.com/ImageMagick/ImageMagick/assets/173410406/2e99f18a-3de7-4778-8feb-741299c70e56)
Granted, is this how it's supposed to be ran?

**Comment by snibgo** [comment, marked as answer]
You should be able to run it from a Windows terminal, in any directory.

Check that tr.exe has been installed, eg:
```
C:\web\im>dir \cygwin64\bin\tr.exe
 Volume in drive C is *****
 Volume Serial Number is ******

 Directory of C:\cygwin64\bin

14/07/2022  20:27            51,219 tr.exe
               1 File(s)         51,219 bytes
```
If tr.exe is there, then add that directory to your system path.

If it isn't there, then you need to select an appropriate package, such as `coreutils`, at the Cygwin Setup screen. That package also contains other useful utilities such as `head` and `tail`.

**Reply by 60horsesinmyherd** [reply]
Okay. Will do. I think now that I've tested it and got it working (albeit in the wrong directory) I can count this problem of mine solved. Thank you @fmw42, and thank you @snibgo!

**Comment by fmw42** [comment]
Here is a script to get the average color in the hexagon cells so that the center pixel will pick it up.  It creates a white hexagon shape on black background and then convolves that with the input image to get the average at every pixel in the input image. So the resulting image could replace the input image in my other script, if you wanted the average color.

```
radius=17.5
diam=`convert xc: -format "%[fx:ceil(2*$radius)]" info:`

# function to create regular polygon vertices
polygonVertices()
	{
	num=$1
	rad1=$2
	rad2=$3
	offset=$4
	ptArr=()
	for ((i=0; i<num; i++)); do
	ang=$((i*360/num + offset))
	xx[$i]=`convert xc: -format "%[fx:$rad2*cos($ang*pi/180)+$rad1-0.5]" info:`
	yy[$i]=`convert xc: -format "%[fx:$rad2*sin($ang*pi/180)+$rad1-0.5]" info:`
	ptArr[$i]="${xx[$i]},${yy[$i]}"
	done
	}

# create hexagon shape image
polygonVertices "6" "$radius" "$radius" "0"
magick -size ${diam}x${diam} xc:black -fill white -draw "polygon ${ptArr[*]}" -alpha off \
-trim +repage -bordercolor black -border 1 tmpS.png

# get dimensions of hexagon shape image
dims=`magick tmpS.png -format "%wx%h" info:`

# create convolution kernel from shape image
kernel=`magick tmpS.png -compress none -depth 8 PGM:- | sed "1d;3d" | sed "1 s/^.*$/$dims:/"`

# apply convolution kernel to input image
magick image.png -define convolve:scale=\! -morphology convolve "$kernel" image_hex_average.png

```

Hexagon shape:

![tmpS](https://github.com/ImageMagick/ImageMagick/assets/13458218/b93107f4-9bb6-4951-8097-2fa7d0c7a814)


Averaged image:

![image_hex_average](https://github.com/ImageMagick/ImageMagick/assets/13458218/ea4f2abc-07d1-4e0a-8924-7062640df7ab)


--------------------------------------

Also I note that the center pixels could be found for every cell from the hexagon image using connected components.

**Comment by snibgo** [comment]
From my new "polyPixB" code, from an image `toes.png` and a polygon image `g3.png`, make the result `g4.png`:

toes.png:
![toes](https://github.com/ImageMagick/ImageMagick/assets/33812368/82ff1a7e-2ce7-4bdd-88ea-65fea42a33fa)

g3.png, which has smaller polygons where toes.png has more detail:
![g3](https://github.com/ImageMagick/ImageMagick/assets/33812368/f74fc3d1-b566-41bc-87a1-29ef47c5e9ba)

output g4.png:
![g4](https://github.com/ImageMagick/ImageMagick/assets/33812368/37dc1183-034b-46cb-a01a-cb9e86dff64b)

**Comment by GeeMack** [comment]
I developed this ImageMagick v7 command, shown here in a simple Windows BAT script, to take any input image and essentially divide it into single color, hexagon shaped cells of 37x31 pixels. 

It takes one argument, either an input image by file name, or an image on the Windows clipboard.  The output image will be the same dimensions as the input.

It basically divides the image into a grid of rectangular cells, resizes each cell to a single 1x1 pixel then scales it back to its starting size and re-assembles the grid. That creates the necessary single color units.

Then two copies of the grid are made and cropped into vertical strips. On one copy, the odd numbered 1, 3, 5... columns are widened and the even 2, 4, 6... columns are narrowed then appended. The other copy has the opposite even/odd columns made wide and narrow.

A mask of tiled hex shaped cells is used to create the transparent, hex shaped areas in the top image, which is then composited over the bottom image to finish making the output with one color hex cells.

A hexagonal, black line grid is composited over that, and the image is completed by cropping it back to its original dimensions. 

    @echo off
    
    :: ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: ::
    :: 
    set OUTFILE=outputfilename.png
    
    if "%~1"=="" ( set INFILE=clipboard: ) else ( set INFILE="%~1" )
    
    magick %INFILE% ^
       +repage ^
       -set option:wxh %%[w]x%%[h] ^
       -background none ^
       -bordercolor none ^
       -write mpr:img0 ^
       -set option:w_div %%[fx:ceil(w/27)+(ceil(w/27)%%2?2:3)] ^
       -set option:h_div %%[fx:ceil(h/31)] ^
       -resize %%[fx:w_div*27]x%%[fx:h_div*31]! ^
       -gravity center ^
       mpr:img0 ^
       -composite ^
       +gravity ^
       -crop %%[w_div]x1@ ^
       -distort affine "0,0 0,%%[fx:(t%%2)*h/h_div/2]" ^
       -crop 1x%%[fx:h_div]@ ^
       -set option:resizer %%[w]x%%[h] ^
       -resize 1x1 ^
       -resize %%[resizer]! ^
       -append ^
       -crop 1x%%[w_div]@ ^
       -distort affine "0,0 0,-%%[fx:(t%%2)*h/h_div/2]" ^
       +append ^
       -write mpr:grid ^
       +delete ^
       ( ^
          -size 54x31 ^
          +antialias ^
          -strokewidth 1 ^
          -stroke black ^
          -fill none ^
          xc:none ^
          -draw "polyline 0,15 10,15 18,0 37,0 45,15  54,15  45,15 37,31 18,31 10,15" ^
          -write mpr:lines ^
          -shave 8x0 ^
          -border 8x0 ^
          -draw "fill black color 27,15 floodfill" ^
          -write mpr:masker ^
          +delete ^
       ) ^
       ( ^
          mpr:grid ^
          -crop 27x%%[h] ^
          -scale %%[fx:(t%%2)?52:2]x%%[h]! ^
          +append ^
       ) ^
       ( ^
          mpr:grid ^
          -crop 27x%%[h] ^
          -scale %%[fx:(t%%2)?2:52]x%%[h]! ^
          +append ^
          -size %%[w]x%%[h] ^
          tile:mpr:masker ^
          -compose copyopacity ^
          -composite ^
          -compose over ^
       ) ^
       -gravity center ^
       -composite ^
       tile:mpr:lines ^
       -composite ^
       -extent %%[wxh] ^
       +gravity ^
          hexerout.png
    
    exit /b
    :: 
    :: ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: ::

Typical input, output examples...
![_ch_hexerinput](https://github.com/ImageMagick/ImageMagick/assets/65475279/392f6b5e-60f2-4489-95ed-e0dc4efa7fcc)
![_ch_hexerout](https://github.com/ImageMagick/ImageMagick/assets/65475279/aa1c61a9-b78b-4607-bd0b-dc791cb7cb1a)
